### ISSUE TITLE
How to get Redux-ORM working with Typescript

### ISSUE THREAD

**Issue by Aryk** [tommikaikkonen/redux-orm#68]
I've already run into a few issues getting redux-orm to work with Typescript. Most of the issues relate to the fact that we are subclassing the Model class from a non-TS project.

In any case, I found the first fix here #29. @tommikaikkonen has since added in the fix for that.

The 2nd issue I encountered is the fact that the static getters on Model are not inherited by Typescript. This took me a while to figure out as it's not immediately obvious.

The way to fix this in the library is to internally not use the getter and setter classes and instead call like getSession() which will call this._session. Then for the users, you could have the static getters defined as usual for easier typing. Not sure if you want to do that, so I had to hack this in.

You can use this as a decorator:

```js
@reduxOrmStaticGetterFix
class User extends Model {

}
```

Here is the code:

```js
// Aryk: even with the 2015 transform classes plugin, (https://babeljs.io/docs/plugins/transform-es2015-classes/)
// it won't do getters and setters, so I have to define them here manually for rexux-orm library.
//
// So I have to add them on here (on the typescript side so that Typescript compiler can find these methods.
// This issue is related to this: https://github.com/Microsoft/TypeScript/issues/1520.
export const reduxOrmStaticGetterFix = (reduxModel) => {

  if (!reduxModel.hasOwnProperty('_sessionData')) {
    Object.defineProperty(reduxModel, '_sessionData', {
      get: function () {
        if (!this.session) return {};
        return this.session.getDataForreduxModel(this.modelName);
      },
    });
  }

  if (!reduxModel.hasOwnProperty('idAttribute')) {
    Object.defineProperty(reduxModel, 'idAttribute', {
      get: function () {
        return this.session.db.describe(this.modelName).idAttribute;
      },
    });
  }

  if (!reduxModel.hasOwnProperty('session')) {
    Object.defineProperty(reduxModel, 'session', {
      get: function () {
        return this._session;
      },
    });
  }

  if (!reduxModel.hasOwnProperty('query')) {
    Object.defineProperty(reduxModel, 'query', {
      get: function () {
        return this.getQuerySet();
      },
    });
  }

  return reduxModel;
};
```

**Comment by Aryk** [comment]
More fixes, in the case you want to subclass QuerySet.

```js
/**
 * Because TS does not implement the ES6 spec: https://github.com/Microsoft/TypeScript/issues/1601, in the library
 * on the `attachQuerySetMethods` method, which calls `forEachSuperClass`, and subsequently
 * `Object.getPrototypeOf(currClass)`, it will not move up the chain and define methods, so calls like Book.filter
 * or Book.first() will not work.
 *
 * So basically, whenever you subclass QuerySet, you must also apply this decorator.
 */
export const reduxOrmSuperClassFixer = (superModelClass) => (modelClass) => {
  modelClass.__proto__ = superModelClass;
  return modelClass;
};

```

**Comment by ch-weiss** [comment]
Is there any news on this? It is really sad that is not possible to use Redux-ORM with TS. You know one thing is whether you are using TS or not the other thing is to make it at least possible for TS users to use Redux-ORM. At the moment it is not possible. Please be so tolerant to make it possible. Thank you.

**Comment by ch-weiss** [comment]
Is there any news on this? It is really sad that is not possible to use redux-rom with TS. Please be so tolerant to make it possible. Thank you.

**Comment by ch-weiss** [comment]
Update: I found a way that seems to work with TS. Here is a link to an example:
https://github.com/didomi/redux-orm-angular/blob/master/test/helpers/models.ts
